File: Graph/[백준 5567] 결혼식.swift
//
//  [백준 5567] 결혼식.swift
//  Algorithm
//
//  Created by 김민 on 4/9/24.
//
// 백준 5567 결혼식 https://www.acmicpc.net/problem/5567

/*
[📌 전략]
 그래프 탐색으로 풀이할 수 있는 문제
 상근이의 친구, 친구의 친구까지만 탐색하기 위한 로직 필요

[✅ 풀이]
 bfs로 풀이하였고, 큐에 넣을 때 (정점, depth)를 enqueue
 상근이의 친구의 친구까지 탐색해야 하므로 depth가 3이상이 되면 더이상 탐색하지 않아도 됨.
*/

let n = Int(readLine()!)!
let m = Int(readLine()!)!
var g = Array(repeating: [Int](), count: n+1)
var visited = Array(repeating: false, count: n+1)
var ans = 0

for _ in 0..<m {
    let line = readLine()!.split(separator: " ").map { Int($0)! }
    let (a, b) = (line[0], line[1])
    g[a].append(b)
    g[b].append(a)
}

func bfs() {
    var q = [(1, 0)] // (정점, depth)
    visited[1] = true

    while !q.isEmpty { // 큐가 빌 때까지
        let (f, cnt) = q.removeFirst()

        for nxt in g[f] { // 인접한 정점 탐색
            if visited[nxt] { continue }
            if cnt + 1 >= 3 { return } // 3 이상일 경우 탐색 중지

            visited[nxt] = true
            ans += 1
            q.append((nxt, cnt+1))
        }
    }
}

bfs()
print(ans)

File: Python/그래프/[백준 5567] 결혼식.py
'''
[백준 5567] 결혼식
https://www.acmicpc.net/problem/5567

[📌 전략]
- 상근이의 친구, 친구와 친구까지의 개수 파악하기
-> 뻗어나가며 그래프 탐색

[✅ 풀이]
1. 양방향 그래프 입력 처리
2. bfs 탐색
    - 큐에는 [현재 노드, 현재 관계 깊이]로 나타냄
    - 큐에 요소가 존재할 동안 로직 반복
    - 첫 번째 요소를 popLeft 했을 때 cnt >= 2인 경우 친구의 친구의 친구가 되는 것이므로 탐색 종료
    - 이외의 경우는 g[cur]의 인접 노드를 순회하며 아직 방문하지 않았을 경우 큐에 (인접 노드, cnt+1)을 넣어 재탐색

[📝 기록]
LV: 실버2
풀이 시간: 22분

dfs로 접근했다가 친구의 친구를 접근하는 과정과 방문 처리하는 과정이 중복이 되는 오류를 발견하여
bfs로 풀이하였음
'''

import sys
from collections import deque

input = sys.stdin.readline

n = int(input())
m = int(input())

g = [[] for _ in range(n+1)]
visited = [False] * (n+1)

# 그래프 입력 처리
for _ in range(m):
    a, b = map(int, input().split())
    g[a].append(b)
    g[b].append(a)

ans = 0

# bfs 탐색
def bfs():
    global ans

    q = deque([[1, 0]])
    visited[1] = True

    while q:
        cur, cnt = q.popleft()

        if cnt >= 2: return

        for nxt in g[cur]:
            if visited[nxt]: continue

            visited[nxt] = True
            ans += 1
            q.append([nxt, cnt+1])

bfs()
print(ans)